Refactoring

itowlson opened the pull request:
This is an extensive refactor to disentangle some of the core functional chunks of the startup and execution pipeline.  It breaks out stages for:

* Populate the cache
* Interpret the routing configuration
* Load Wasm modules
* Build routing table
* Serve requests

A general theme is increased up front parsing (in the most general sense).  For example, instead of using the suffix whenever we need to determine if a route is prefix or exact, the PR parses the route once into a `RoutePattern` enum, meaning that URL matching tests can be written using a clear (well, clearer to me) match pattern.

In support of the refactor the PR also introduces a number of new tests, the main set of which is the "end-to-end-ish" tests in `lib.rs`.  These don't go through the full HTTP stack but do pass `hyper::Request` objects in at the routing table level so they pass through the full route matching/EV generation/response composition pipeline.  There is plenty more we could do here though!

Finally, I've tried to improve some of the error reporting by adding context to the errors where we raise them.  This is pretty patchy so far but I'll continue to work on it!

There is still work to be done in improving the file-level organisation but hopefully readers and maintainers will find this clearer and more robust overall.

radu-matei commented on Cargo.toml:
```diff
@@ -35,3 +35,4 @@ wasmtime                        = "0.30"
 wasmtime-wasi                   = "0.30"
 wasmtime-cache                  = "0.30"
 wat                             = "1.0.37"
+chrono = "0.4.19"
```
nit: arrange TOML entries to the right.

itowlson replied:
NOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO

radu-matei commented on src/bindle_util.rs:
```diff
@@ -0,0 +1,381 @@
+use bindle::{Invoice, Parcel};
+
+// TODO: this file is a bit of a cop-out but will be useful during
+// the transition.  Find better homes for these things!
```
Is this file still to be used as-is?

itowlson replied:
Hmm, probably.  The structure isn't great but I think the concepts are mostly right and the organisation can can be refined over time.  I'll take another look but may end up just deleting the comment.  Thanks for the catch.

radu-matei commented on src/emplacer.rs:
```diff
@@ -0,0 +1,288 @@
+        }.with_context(|| "Error caching assets from bindle")
+    }
+
+    // TODO: NO! NO! NO!
```
"NO! NO! NO!"

itowlson replied:
I haven't found a better approach yet though _sigh_

radu-matei commented on src/emplacer.rs:
```diff
@@ -0,0 +1,288 @@
+    }
+
+    // TODO: do not like having bindle specifics here
+    pub async fn read_invoice(&self, invoice_id: &bindle::Id) -> anyhow::Result<bindle::Invoice> {
```
If we do have a `bindle_utils` module, would this be better placed there?

itowlson replied:
No, this refers specifically to reading through the emplacer.

radu-matei commented on src/emplacer.rs:
```diff
@@ -0,0 +1,288 @@
+        Ok(())
+    }
+
+    async fn emplace_as_assets(&self, reader: &impl BindleReader, invoice_id: &bindle::Id, parcels: &Vec<bindle::Parcel>) -> anyhow::Result<()> {
```
Given the struct is already an "Emplacer", what do you think about simplifying the names of all the methods that also contain "emplace_"?

itowlson replied:
Open to suggestions but just removing the `emplace` prefix would reduce clarity.

radu-matei commented on src/emplacer.rs:
```diff
@@ -0,0 +1,288 @@
+    
+}
+
+pub(crate) fn invoice_cache_key(id: &bindle::Id) -> String {
```
Same re: being part of a `bindle_utils` module.

itowlson replied:
This is specific to the way the emplacer organises the file cache.  It's not a generic bindle helper.  However, it should be private to the emplacer module - it was temporarily public because it was temporarily shared with the old code.  Thanks!

radu-matei commented on src/header_util.rs:
```diff
@@ -0,0 +1 @@
+
```
Is this supposed to be committed?

itowlson replied:
Ugh.  Guessing that this got created, then deleted, then somehow managed to recreate itself during a hairy rebase.  Thanks for the catch.

itowlson commented:
@technosophos Per our chat:

1. Fixed it so the health check route takes precedence over user routes (and added a test)
2. Confirmed the WASM cache file is used (in `wasm_runner::new_store_and_engine`)

technosophos commented on src/dispatcher.rs:
```diff
@@ -0,0 +1,378 @@
+    /// - route
+    /// - host
+    //
+    // TODO: this ^^ is the original comment on Module::id() but it doesn't seem to
```
You are correct. I removed the host part when I removed virtual hosting a few months ago. Missed the comment text.

technosophos commented on src/wasm_module.rs:
```diff
@@ -0,0 +1,35 @@
+// In future this might be pre-instantiated or something like that, so we will
+// just abstract it to be safe.
+#[derive(Clone, Debug)]
+pub enum WasmModuleSource {
+    Blob(Arc<Vec<u8>>),
+}
```
@radu-matei Does it kinda make sense how you could add something here to add support for pre-inited modules?

technosophos commented on src/dispatcher.rs:
```diff
@@ -0,0 +1,378 @@
+}
+
+impl RoutingTable {
+    pub fn build(source: &LoadedHandlerConfiguration, global_context: RequestGlobalContext) -> anyhow::Result<RoutingTable> {
```
@radu-matei I already chatted with @itowlson about this when we paired to code review, but the pre-caching of modules is now gone(-ish) from the code. We are not sure whether to put that back in, or two switch to the pre-init stuff and plug that in here or at the top `main.rs` level.

technosophos replied:
I take back what I said about the cache. See https://github.com/deislabs/wagi/pull/125#issuecomment-933987027

technosophos reviewed (approved):
Assuming the few nits noted are taken care of, I am totally good with merging this refactor. It is excellent, and I am really thrilled (chuffed, one might say) to see Wagi take on a more polished look.
